### PR TITLE
editoast: add macros schemas! and routes!

### LINKED ISSUE #4994: editoast+openApi: automatedly generated openApi file is not valid
### What happened?

OpenApi is broken and cannot be used to generate types for the frontend anymore.

This is likely because of the automatic generation of TimetableImportPathLocation, because "Discriminators require references, not inline schemas".
We should either modify the automated generation of the OpenApi or write the description of the concerned endpoint manually (and not automatedly).

### What did you expect to happen?

The automated generation of the openApi should generate a valid openApi file.

### How can we reproduce it (as minimally and precisely as possible)?

1. cd front
2. yarn generate-types

### What operating system, browser and environment are you using?

- **Browser**: Firefox v109
- **OS**: Nixos v22.11
- **Env**: Local

### OSRD version (top right corner `Account` button > `Informations`)

e0ad0648

### PR DESCRIPTION
**Pull request by leovalais**
Part of #3625, to merge before #4802

* macro `routes!` to replace `actix::services!` and `actix::web::scope` + endpoint listing in views::OpenApiRoot
* macro `schemas!` that does the same for schemas
* deletes the `discriminator:` property because it conflicts with RTK
* also adds the crate `paste` for ident interpolation

### CODE REVIEW

File: editoast/src/views/mod.rs [hunk 1]
```diff
@@ -65,25 +79,89 @@ pub fn study_routes() -> impl HttpServiceFactory {
...
-            timetable::TimetableImportTrain
-        ),
-        responses()
-    )
+    paths(),
+    components(responses())
 )]
 pub struct OpenApiRoot;
 
+fn remove_discriminator(schema: &mut RefOr<Schema>) {
```
Comment (flomonster) [inline editoast/src/views/mod.rs]: We could move this to the `openapi` module.

File: editoast/src/views/mod.rs [hunk 2]
```diff
@@ -65,25 +79,89 @@ pub fn study_routes() -> impl HttpServiceFactory {
...
+                        remove_discriminator(&mut content.schema);
+                    }
+                }
+            }
+        }
+        if let Some(components) = openapi.components.as_mut() {
+            for component in components.schemas.values_mut() {
+                remove_discriminator(component);
+            }
+        }
```
Comment (flomonster) [inline editoast/src/views/mod.rs]: We could make a function for this

File: editoast/src/views/mod.rs [hunk 3]
```diff
@@ -16,25 +16,38 @@ pub mod study;
...
 use redis::cmd;
 use serde_derive::{Deserialize, Serialize};
-use utoipa::{OpenApi, ToSchema};
+use utoipa::openapi::{Array, Object};
+use utoipa::{
+    openapi::{schema::AnyOf, AllOf, OneOf, RefOr, Schema},
+    OpenApi, ToSchema,
+};
+
+fn routes_v2() -> Routes<impl HttpServiceFactory> {
```
Comment (flomonster) [inline editoast/src/views/mod.rs]: We could add a comment to explain that this is temporarily used for the migration to utoipa.

File: editoast/src/views/mod.rs [hunk 4]
```diff
@@ -16,25 +16,38 @@ pub mod study;
...
+    crate::routes! {
+        health,
+        version,
+        core_version,
+    }
```
Comment (flomonster) [inline editoast/src/views/mod.rs]: Shouldn't we call `timetable::routes()` here?

**Review by flomonster — CHANGES_REQUESTED** [review]
Thanks for this PR.
I made a few suggestions.
I'm not a declarative macro expert so I hope I didn't miss something.

**Review by flomonster — APPROVED** [review]

**Review by Akctarus — APPROVED** [review]
LGTM

File: front/src/modules/conflict/components/ConflictsList.tsx [hunk 5]
```diff
@@ -6,9 +6,9 @@ import { sec2time } from 'utils/timeManipulation';
 
 export type Conflict = {
```
Comment (clarani) [inline front/src/modules/conflict/components/ConflictsList.tsx]: Can you remove this type and use the new one from osrdEditoastApi ? :) Also you should use the new type in Timetable.tsx too

**Review by clarani — CHANGES_REQUESTED** [review]

**Review by clarani — APPROVED** [review]
LGTM, tested in local ! :heavy_check_mark:

**Pull request merged**